Remove "ech_nonce" extension and clarify client behavior (#292)
tlswg/draft-ietf-tls-esni: draft-ietf-tls-esni.md

## Sending an encrypted ClientHello {#send-ech}

To offer ECH, the client first chooses a suitable ECH configuration. To
determine if a given `ECHConfig` is suitable, it checks that it supports the KEM
algorithm identified by `ECHConfig.kem_id` and at least one KDF/AEAD algorithm
identified by `ECHConfig.cipher_suites`. Once a suitable configuration is found,
the client selects the cipher suite it will use for encryption. It MUST NOT
choose a cipher suite not advertised by the configuration.

Next, the client constructs the ClientHelloOuter message just as it does a
standard ClientHello, with the exception of the following rules:

> 1. It MUST offer to negotiate TLS 1.3.

Reviewer: TLS 1.3 or above.

Author: Done.

> 1. It MUST offer to negotiate TLS 1.3.

Reviewer 2: ```suggestion
1. It MUST offer to negotiate TLS 1.3 or above.
```

Author: Done.

1. It MUST include an "encrypted_client_hello" extension with a payload
   constructed as described below.
1. The value of `ECHConfig.public_name` MUST be placed in the "server_name"
   extension.

> 1. It SHOULD offer dummy values for all other sensitive extensions, including

Author: Done. I also think we should offer this guidance. I'm leaving doing so as an "OPEN ISSUE".

   [[TODO: List the known sensitive extensions.]]

> 1. [[OPEN ISSUE: Any other restrictions on the ClientHelloOuter?]]

Author: Done. The text needs to be updated to reflect this new rule, so I'm adding an "OPEN ISSUE" for this here. I'd be happy to do this as part of this PR, but the change will require additional review.

## Sending an encrypted ClientHello {#send-ech}

To offer ECH, the client first chooses a suitable ECH configuration. To
determine if a given `ECHConfig` is suitable, it checks that it supports the KEM
algorithm identified by `ECHConfig.kem_id` and at least one KDF/AEAD algorithm
identified by `ECHConfig.cipher_suites`. Once a suitable configuration is found,
the client selects the cipher suite it will use for encryption. It MUST NOT
choose a cipher suite not advertised by the configuration.

Next, the client constructs the ClientHelloOuter message just as it does a
standard ClientHello, with the exception of the following rules:

1. It MUST offer to negotiate TLS 1.3 or above.
1. It MUST include an "encrypted_client_hello" extension with a payload
   constructed as described below.
1. The value of `ECHConfig.public_name` MUST be placed in the "server_name"
   extension.

> 1. It MUST NOT include a PSK or PSK binder. [[OPEN ISSUE: The text needs to be
>    updated to reflect this rule.]]

Reviewer 2: ```suggestion
1. It MUST NOT include a "pre_shared_key" extension. [[OPEN ISSUE: The text needs to be
   updated to reflect this rule.]]
```

Author: Excellent! Done.

The client then constructs the ClientHelloInner message just as it does a
standard ClientHello, with the exception of the following rules:

1. It MUST NOT offer to negotiate TLS 1.2 or below.
1. It MUST NOT offer to resume any non-ECH PSK or any session for TLS 1.2 and
   below.
1. It MAY offer any other extension in the ClientHelloOuter except those that
   have been incorporated into the ClientHelloInner as described in
   {{outer-extensions}}.
1. It MAY copy any other field from the ClientHelloOuter except
   ClientHelloOuter.random. Instead, It MUST generate a fresh
   ClientHelloInner.random using a secure random number generator. (See
   {{flow-client-reaction}}.)
1. It SHOULD contain TLS padding {{!RFC7685}} as described in {{padding}}.

The client might duplicate non-sensitive extensions in both messages. However,
implementations need to take care to ensure that sensitive extensions are not

> offered in the CllientHelloOuter. [[OPEN ISSUE: We should provide guidance on

Reviewer 2: ```suggestion
offered in the ClientHelloOuter. [[OPEN ISSUE: We should provide guidance on
```

Author: Duh. Done.

> what extensions are sensitive and suggest suitable substitutes.]]

Reviewer 2: Why substitutes? Why not just drop them from ClientHelloOuter entirely?

Author: "substitutes" = "innocuous values for sensitive extensions". Is that more clear?

## Client-Facing Server

Upon receiving an "encrypted_client_hello" extension, the client-facing server
MUST check that it is able to negotiate TLS 1.3 or greater. If not, it MUST
abort the connection with a "handshake_failure" alert.

The ClientECH value is said to match a known ECHConfig if there exists
an ECHConfig that can be used to successfully decrypt
ClientECH.encrypted_ch. This matching procedure should be done using
one of the following two checks:

1. Compare ClientECH.config_id against identifiers of known ECHConfig
   and choose the one that matches.
2. Use trial decryption of ClientECH.encrypted_ch with known ECHConfig
   and choose the one that succeeds.

Some uses of ECH, such as local discovery mode, may omit the ClientECH.config_id
since it can be used as a tracking vector. In such cases, trial decryption
should be used for matching ClientECH to known ECHConfig. Unless specified by
the application using (D)TLS or externally configured on both sides,
implementations MUST use the first method.

If the ClientECH value does not match any known ECHConfig structure, it
MUST ignore the extension and proceed with the connection, with the following
added behavior:

> - It SHOULD include the "encrypted_client_hello" extension in its

Reviewer 2: Please revert this change -- this MUST was there so clients could rely on this particular version of ECH being deployable. (We can discuss it in a separate issue.)

```suggestion
- It MUST include the "encrypted_client_hello" extension in its
```

Author: Ack. Fixing in #287, then rebasing here.

Note that an unrecognized ClientECH.config_id value may be a GREASE ECH
extension (see {{grease-extensions}}), so it is necessary for servers to proceed
with the connection and rely on the client to abort if ECH was required. In
particular, the unrecognized value alone does not indicate a misconfigured ECH
advertisement ({{misconfiguration}}). Instead, servers can measure occurrences
of the "ech_required" alert to detect this case.

If the ClientECH value matches a known ECHConfig, the server then
decrypts ClientECH.encrypted_ch, using the private key skR corresponding
to ECHConfig, as follows:

~~~
context = SetupBaseR(ClientECH.enc, skR, "tls13-ech")

ClientHelloInner =
  context.Open("", ClientECH.encrypted_ch)

ech_hrr_key = context.Export("tls13-ech-hrr-key", 16)
~~~

If decryption fails, the server MUST abort the connection with a "decrypt_error"
alert. Next, the server MUST scan ClientHelloInner for any "outer_extension"
extensions and substitute their values with the values in ClientHelloOuter. It
MUST first verify that the hash found in the extension matches the hash of the
extension to be interpolated in and if it does not, abort the connections with a

> "decrypt_error" alert.

Reviewer 2: Should the server also assert that the random values are different?

Author: What would be the purpose?

Reviewer 2: (Chatted offline -- maybe not now!)

be ignored by all other servers in the deployment when not negotiating ECH,
including servers which do not implement this specification.

## Backend Server Behavior {#backend-server-behavior}

When the client-facing server accepts ECH, it forwards the ClientHelloInner to
the backend server, who terminates the connection. If the ClientHelloInner
contains an empty "encrypted_client_hello" extension, then the backend server
MUST confirm ECH acceptance by setting ServerHello.random[24:32] to

~~~~
    accept_confirmation = HKDF-Expand-Label(
        HKDF-Extract(0, ClientHelloInner.random),
        "ech-eccept-confirmation",
        ServerHello.random[0:24], 8)
~~~~

where HKDF-Expand-Label and HKDF-Extract are as defined in {{RFC8446}}. The
value of ServerHello.random[0:24] is generated as usual by invoking a secure

> pseudorandom number generator.

Reviewer 2: This doesn't seem part of this PR?

Author: This PR is based on #287, which is why this text appears here. If you focus your review on the last commit, it should be clear what the changes are.

To signal acceptance, the backend server overwrites 8 bytes of its
ServerHello.random with a value derived from the ClientHelloInner.random. (See
{{backend-server-behavior}} for details.) In the presence of an active
adversary, this behavior increases the likelihood of the ServerHello.random
colliding with the ServerHello.random of a previous session, potentially
reducing the overall security of the protocol. However, the remaining 24 bytes
provide enough entropy to ensure this is not a practical avenue of attack.

On the other hand, the probability that two 8-byte strings are the same is
non-negligible. This poses a modest operational risk. Suppose the client-facing
server terminates the connection (i.e., ECH is rejected or bypassed): if the
last 8 bytes of its ServerHello.random coincide with the confirmation signal,
then the client will incorrectly presume acceptance and proceed as if the
backend server terminated the connection.

Note that the same bytes of the ServerHello.random are used to signal to the
client that an earlier version of TLS was negotiated (see {{RFC8446}}, Section
4.1.3). The backend server's signal of acceptance does not interfere with this

> mechanism because ECH is only supported in TLS 1.3 or higher.

Reviewer 2: Ditto here on this being unrelated to the nonce?

~~~
{: #flow-diagram-client-reaction title="Client reaction attack"}

> The ClientHelloInner.random prevents this attack. In particular, since the

Reviewer 2: ```suggestion
ClientHelloInner.random prevents this attack. In particular, since the
```

Author: Done.

1. encrypted_client_hello(0xff08), with "TLS 1.3" column values being set to
   "CH, EE", and "Recommended" column being set to "Yes".

> 3. outer_extension(0xff04), with the "TLS 1.3" column values being set to "CH",

Reviewer 2: Let's move this to 0xff04.

Author: What should we do about "outer_extension" (currently 0xff04)?

Reviewer 2: Oops, I meant 0xff03. Heh. Let's just revisit these values in a followup PR.
